### ISSUE TITLE
Potential node crash when doing base-view pairing after decreasing RF

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylladb#21492]
The base-view pairing assumes base and view replica sets within a DC are always of same size, but turns this assumption can be broken when decreasing RF, since the leaving replicas of base and view will go through independent migrations.

This assumption is explicitly defined in get_view_natural_endpoint() with this assert: `SCYLLA_ASSERT(base_endpoints.size() == view_endpoints.size());`

    
Assume you have a base and view with RF=2

```
 table_name | last_token          | replicas
------------+---------------------+----------------------------------------------------------------------------------------
      users | 9223372036854775807 | [(9ac667ad-5d5a-4e26-b70d-483043f58eb0, 0), (3993b4c5-9581-4bc5-9dec-775a9ee032c3, 0)]
   users_mv | 9223372036854775807 | [(3993b4c5-9581-4bc5-9dec-775a9ee032c3, 1), (9ac667ad-5d5a-4e26-b70d-483043f58eb0, 1)]
```
   
   
When decreasing RF to 1, the replica sets can be found in this state

```
 table_name | last_token          | replicas
------------+---------------------+----------------------------------------------------------------------------------------
      users | 9223372036854775807 | [(9ac667ad-5d5a-4e26-b70d-483043f58eb0, 0), (3993b4c5-9581-4bc5-9dec-775a9ee032c3, 0)]
   users_mv | 9223372036854775807 |                                            [(3993b4c5-9581-4bc5-9dec-775a9ee032c3, 1)]
```

That happened because system was able to complete migration for removing view replica, before the one for the base.

If migration for view completes while the migration for base is still at write_both_read_new or earlier (points to the current replica set, not next), the assumption is broken during pairing and node crashes.
```
INFO  2024-11-07 16:46:12,553 [shard 0:stmt] view - base_endpoints=2 ([9ac667ad-5d5a-4e26-b70d-483043f58eb0, 3993b4c5-9581-4bc5-9dec-775a9ee032c3]), view_endpoints=1 ([3993b4c5-9581-4bc5-9dec-775a9ee032c3])
scylla: db/view/view.cc:1779: std::optional<gms::inet_address> db::view::get_view_natural_endpoint(const locator::effective_replication_map_ptr &, const locator::effective_replication_map_ptr &, bool, const dht::token &, const dht::token &, bool, replica::cf_stats &): Assertion `base_endpoints.size() == view_endpoints.size()' failed.
Aborting on shard 0, in scheduling group statement.
Backtrace:
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x41e144
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x3e2c85
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x410ae8
  /lib64/libc.so.6+0x40cff
  /lib64/libc.so.6+0x99663
  /lib64/libc.so.6+0x40c4d
  /lib64/libc.so.6+0x28901
  /lib64/libc.so.6+0x2881d
```

If the system finds itself in this scenario, this can easily escalate into a serious outage since replicas doing view updates will share the same tablet metadata state.

A possible solution is to exclude leaving replica (without a corresponding pending) from doing view updates.

**Comment by kostja** [comment]
@bitpathfinder afaiu you're looking into actual data movements during RF decrease right now, can you comment? How will it work when your patch is in?
cc @kbr-scylla

**Comment by bitpathfinder** [comment]
> [@bitpathfinder](https://github.com/bitpathfinder) afaiu you're looking into actual data movements during RF decrease right now, can you comment? How will it work when your patch is in? cc [@kbr-scylla](https://github.com/kbr-scylla)

Looks like changes made in my PR (https://github.com/scylladb/scylladb/pull/21289) won't help here, since these are changing the replication factor calculation, not the replicas set.

**Comment by bhalevy** [comment]
I'll address that in https://github.com/scylladb/scylladb/pull/21453

**Comment by raphaelsc** [comment]
> I'll address that in [#21453](https://github.com/scylladb/scylladb/pull/21453)

Could you please share how you plan to address it?

**Comment by raphaelsc** [comment]
> > > I'll address that in [#21453](https://github.com/scylladb/scylladb/pull/21453)
> > 
> > 
> > Could you please share how you plan to address it?
> 
> `get_view_natural_endpoint` is allowed to return nullptr is it can't find a pair for any reason. So if there are too few view replicas than base replicas we can just return nullptr and get away with it. That doesn't fix the underlying issue of desynchronized tablet deallocation when reducing rf, but it should prevent the undefined behavior.
> 
> preview:
> 
>     auto base_it = std::find(base_endpoints.begin(), base_endpoints.end(), me);
>     if (base_it == base_endpoints.end()) {
>         // This node is not a base replica of this key, so we return empty
>         // FIXME: This case shouldn't happen, and if it happens, a view update
>         // would be lost.
>         ++cf_stats.total_view_updates_on_wrong_node;
>         return {};
>     }
>     size_t idx = base_it - base_endpoints.begin();
>     if (idx >= view_endpoints.size()) {
>         // There are fewer view replicas than base replicas
>         // FIXME: This might still happen when reducing replication factor with tablets,
>         // see https://github.com/scylladb/scylladb/issues/21492
>         ++cf_stats.total_view_updates_failed_pairing;
>         return std::nullopt;
>     }
>     return resolve(view_endpoints[idx]);

I don't think this is a good fix, since it will cause all view updates to be lost (from all base replicas) while the system finds itself in this state. we should avoid the crash of course, but this doesn't really fix the problem.

**Comment by wmitros** [comment]
I managed to replicate this crash but the scenario that I observed seems a bit different than what's described here. My scenario is as follows:
1. I created a table with a materialized view and started writing to it
2. I started altering the repliaction factor of the table's keyspace back-and-forth between 2 and 1
3.  The node crashed on the `base_endpoints.size() == view_endpoints.size()` assert. The difference is, the node crashed while increasing the RF.
4. After running this scenario a couple times I only got the crash in this case, not when decreasing RF.

@raphaelsc could you describe how did you encounter this crash? We may have issues in both cases, but perhaps it's actually only increasing the RF that causes the issue - the logs were a bit misleading in my case, the last log describing the replication factor change was `alter_keyspace - Using Replication Factor datacenter1=1` and I could only see that it was the other change by increasing the log level/looking at timing.

The fix for this shouldn't be that complicated:

When `base_endpoints.size() > view_endpoints.size()`, the fix for either case might be the same - considering that during rf change, we only add/remove the last replica (referring to the affected dc) in the replica list, if we encounter this problem we can simply filter out the last replicas of datacenters where the replica count doesn't match for base and view. I might be wrong but the new view replica should get the update anyway because it should be pending at the time

I believe we can also get the `base_endpoints.size() < view_endpoints.size()` case. We shouldn't just remove the view endpoint from the replica list because it won't get the update in this scenario. Instead, we should treat the last view replicas (also in dc where base and view count don't match) as a pending replica, so that it gets the update from some node.

One more comment regarding the decreasing RF scenario: from the logs, it looks very unlikely we desync the replicas in this case. We start both (base and view) tablet migrations of the leaving replicas in the same topology update and we skip streaming in this case (as there are no new replicas), so if nothing goes wrong, both replica sets should get updated at the same time

**Comment by raphaelsc** [comment]
> I managed to replicate this crash but the scenario that I observed seems a bit different than what's described here. My scenario is as follows:
> 
> 1. I created a table with a materialized view and started writing to it
> 2. I started altering the repliaction factor of the table's keyspace back-and-forth between 2 and 1
> 3. The node crashed on the `base_endpoints.size() == view_endpoints.size()` assert. The difference is, the node crashed while increasing the RF.
> 4. After running this scenario a couple times I only got the crash in this case, not when decreasing RF.
> 
> [@raphaelsc](https://github.com/raphaelsc) could you describe how did you encounter this crash? We may have issues in both cases, but perhaps it's actually only increasing the RF that causes the issue - the logs were a bit misleading in my case, the last log describing the replication factor change was `alter_keyspace - Using Replication Factor datacenter1=1` and I could only see that it was the other change by increasing the log level/looking at timing.

I think it can possibly be reproduced both ways. To reproduce it when reducing RF, it requires the coordinator completing first the migration for view replica, while the migration for base is still in early stages. Very hard to reproduce indeed. Maybe you can inject some artificial delay somewhere (perhaps in storage_service::stream_tablet()) for an unit test. Were you able to reproduce it when increasing RF on master? I remember @ptrsmrn had some patch to allocate atomically new replicas for base and view when increasing RF. The thing with decreasing RF is that removal of base and view replicas aren't synchronized, since removal is migration based.

**Comment by raphaelsc** [comment]
To reproduce it, I had to increase RF to 3, and back to 2, in a loop, while doing writes to base in parallel. Rare were the cases where base and view had a mismatch in replica count.

**Comment by wmitros** [comment]
> > [@raphaelsc](https://github.com/raphaelsc) could you describe how did you encounter this crash? We may have issues in both cases, but perhaps it's actually only increasing the RF that causes the issue - the logs were a bit misleading in my case, the last log describing the replication factor change was `alter_keyspace - Using Replication Factor datacenter1=1` and I could only see that it was the other change by increasing the log level/looking at timing.
> 
> I think it can possibly be reproduced both ways. To reproduce it when reducing RF, it requires the coordinator completing first the migration for view replica, while the migration for base is still in early stages. Very hard to reproduce indeed. Maybe you can inject some artificial delay somewhere (perhaps in topology_coordinator::handle_tablet_migration()) for an unit test. Were you able to reproduce it when increasing RF on master? I remember [@ptrsmrn](https://github.com/ptrsmrn) had some patch to allocate atomically new replicas for base and view when increasing RF. The thing with decreasing RF is that removal of base and view replicas aren't synchronized, since removal is migration based.

Yes, I was running my test on master. Contrary to https://opensource.docs.scylladb.com/stable/kb/rf-increase.html, it looks like we also do streaming when increasing RF, in particular Scylla crashed quickly after logging
```
DEBUG 2024-11-18 03:27:27,105 [shard 0: gms] raft_topology - Will set tablet b1ee4ae0-a552-11ef-886c-17b6b6ba78d4:0 stage to use_new
DEBUG 2024-11-18 03:27:27,105 [shard 0: gms] raft_topology - Tablet b43072b0-a552-11ef-886c-17b6b6ba78d4:0 still doing streaming
```
In contrast, while performing tablet migrations during RF reduction, we go through the following stages there:
```
DEBUG 2024-11-18 03:27:03,631 [shard 0: gms] raft_topology - Will set tablet b1ee4ae0-a552-11ef-886c-17b6b6ba78d4:0 stage to write_both_read_old
DEBUG 2024-11-18 03:27:03,631 [shard 0: gms] raft_topology - Will set tablet b43072b0-a552-11ef-886c-17b6b6ba78d4:0 stage to write_both_read_old
```
```
DEBUG 2024-11-18 03:27:03,669 [shard 0: gms] raft_topology - Will set tablet b1ee4ae0-a552-11ef-886c-17b6b6ba78d4:0 stage to streaming
DEBUG 2024-11-18 03:27:03,669 [shard 0: gms] raft_topology - Will set tablet b43072b0-a552-11ef-886c-17b6b6ba78d4:0 stage to streaming
```
```
INFO  2024-11-18 03:27:03,682 [shard 0: gms] raft_topology - Skipped tablet streaming (rebuild) of b43072b0-a552-11ef-886c-17b6b6ba78d4:0 as no pending replica found
DEBUG 2024-11-18 03:27:03,682 [shard 0: gms] raft_topology - streaming for tablet b43072b0-a552-11ef-886c-17b6b6ba78d4:0 resolved.
INFO  2024-11-18 03:27:03,682 [shard 0: gms] raft_topology - Skipped tablet streaming (rebuild) of b1ee4ae0-a552-11ef-886c-17b6b6ba78d4:0 as no pending replica found
DEBUG 2024-11-18 03:27:03,682 [shard 0: gms] raft_topology - streaming for tablet b1ee4ae0-a552-11ef-886c-17b6b6ba78d4:0 resolved.
```
```
DEBUG 2024-11-18 03:27:03,684 [shard 0: gms] raft_topology - Will set tablet b43072b0-a552-11ef-886c-17b6b6ba78d4:0 stage to write_both_read_new
DEBUG 2024-11-18 03:27:03,684 [shard 0: gms] raft_topology - Will set tablet b1ee4ae0-a552-11ef-886c-17b6b6ba78d4:0 stage to write_both_read_new
```
```
DEBUG 2024-11-18 03:27:03,732 [shard 0: gms] raft_topology - Will set tablet b1ee4ae0-a552-11ef-886c-17b6b6ba78d4:0 stage to use_new
DEBUG 2024-11-18 03:27:03,732 [shard 0: gms] raft_topology - Will set tablet b43072b0-a552-11ef-886c-17b6b6ba78d4:0 stage to use_new
```
From what I'm seeing, each of these pairs of updates is prepared in one corresponding `handle_tablet_migration()` call, which should apply all prepared group0 mutations atomically in a `update_topology_state` call. Because we're skipping streaming in this case, it doesn't look like we can encounter the state that I posted for the `RF increase` case, but perhaps there's some other scenario where one of the base/view can get delayed before reaching `use_new` and they can get desynced
